2022/src/day13.rs

```diff
@@ -0,0 +1,191 @@
+use crate::day13::Value::List;
+use crate::puzzle::{Puzzle, PuzzleFn::I32};
+use itertools::Itertools;
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::i32 as nom_i32;
+use nom::combinator::map as nom_map;
+use nom::multi::separated_list0;
+use nom::sequence::delimited;
+use nom::IResult;
+use std::cmp::Ordering;
+use Value::Int;
+
+pub(crate) const PUZZLE: Puzzle = Puzzle {
+    day: 13,
+    part1: I32(part1),
+    part2: I32(part2),
+};
+
+#[derive(Clone, Debug)]
+enum Value {
+    Int(i32),
+    List(Vec<Value>),
+}
+
+impl Eq for Value {}
+
+impl PartialEq<Self> for Value {
+    fn eq(&self, other: &Self) -> bool {
+        match self {
+            Int(left) => match other {
+                Int(right) => left == right,
+                List(_) => List(vec![self.clone()]) == *other,
+            },
+            List(left_list) => match other {
+                Int(_) => self == &List(vec![other.clone()]),
+                List(right_list) => left_list == right_list,
+            },
+        }
+    }
+}
+
+impl PartialOrd<Self> for Value {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Value {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self {
+            Int(left) => match other {
+                Int(right) => left.cmp(right),
+                List(_) => List(vec![self.clone()]).cmp(other),
+            },
+            List(left_list) => match other {
+                Int(_) => self.cmp(&List(vec![other.clone()])),
+                List(right_list) => left_list.cmp(right_list),
+            },
+        }
+    }
+}
+
+fn parse_input(input: &str) -> Vec<(Value, Value)> {
+    input.split("\n\n").map(parse_packets).collect()
+}
+
+fn parse_packets(i: &str) -> (Value, Value) {
+    let mut lines = i.lines();
+    let (_, left) = list(lines.next().unwrap()).unwrap();
+    let (_, right) = list(lines.next().unwrap()).unwrap();
+    (left, right)
```

Comment:
wdyt of this style vs  let (left,right) = ( foo.unwrap().1, foo.unwrap().1); ?

Reply:
The former is a little easier to debug maybe? For short statements I like the later though